### PR TITLE
Fix StatisticsLogger Deadlock

### PR DESCRIPTION
**Pull request by DieHertz**
For a detailed description have a look at #1329.

Review and suggestions are welcome, in case I've missed something.

### CODE REVIEW

**Comment by DieHertz** [comment]
Seems like in this variant it also causes a more obvious, but a deadlock :-(

**Comment by DieHertz** [comment]
When `StatisticsLogger` is initialized from `subscribe` thread, `subs_mutex_` gets locked twice in the sequence `subscribe (topic) -> advertise (statistics)`. `subs_mutex_` could be made a `recursive_mutex` to allow locking it more than once.

Also, `StatisticsLogger` shouldn't be initialized for topics `/clock` and `/statistics` to avoid locking inside `ros::init`.

Another solution (which works in my tests, doesn't reproduce the original issue, but I can't be sure it's entirely correct yet) is to introduce a `advertise_subscribe_mutex_` into `TopicManager` and lock it inside `advertise/unadvertise/subscribe/unsubscribe`.

Which one should I go with? Need your advice.

**Comment by DieHertz** [comment]
Decided to go with another approach, extracting `StatisticLogger::pub_` initialization to `TopicManager::subscribe`, but outside of `Subscription` creation.
That way I can no longer reproduce any of the issues I had nor other associated deadlock that were possible.

File: clients/roscpp/include/ros/statistics.h [hunk 1]
```diff
@@ -84,6 +87,7 @@ class ROSCPP_DECL StatisticsLogger
   double pub_frequency_;
 
   // publisher for statistics data
+  boost::atomic<bool> pub_initialized_;
```
Comment (kev-the-dev) [inline clients/roscpp/include/ros/statistics.h]: Above comment is no longer accurate.

**Review by kev-the-dev — COMMENTED** [review]
EDIT: See the issue for the backtrace and code insights what show the deadlock scenario

I've been unable to reproduce this on my machine with the following program like the one you mentioned in the issue (I've run dozens of times). Still looking into it... Do you have a  specific program I can see this with/a rostest which produces this?:
```
#include <ros/ros.h>
#include <std_msgs/Int32.h>
#include <iostream>
#include <vector>

void c1(const std_msgs::Int32::ConstPtr& msg)
{
  std::cout << "C1" << std::endl;
}

void c2(const std_msgs::Int32::ConstPtr& msg)
{
  std::cout << "C2" << std::endl;
}

int main(int argc, char** argv) {
  ros::init(argc, argv, "statistics_logger_deadlock");
  ros::NodeHandle nh;
  
  std::vector<ros::Subscriber*> subs;
  // This call returns properly
  for (size_t i = 0; i < 100; ++i)
  {
    ros::Subscriber* sub = new ros::Subscriber();
    *sub = nh.subscribe<std_msgs::Int32>("/topic", 1, c1);
    ros::spinOnce();
    subs.push_back(sub);
  }

  std::cout << "Bout to spin" << std::endl;

  ros::spin();
}
```

**Comment by DieHertz** [comment]
To reproduce it I have run a really CPU intensive computation task on the same machine.
Visual inspection of the code and a captures trace from 2 locked threads shows that the situation is indeed possible, albeit not too frequent.

I'll try to come up with a minimal project to reproduce this weekend though.

File: clients/roscpp/include/ros/statistics.h (hunk 1 above)
Comment (DieHertz) [reply clients/roscpp/include/ros/statistics.h]: Will swap the lines

**Comment by kev-the-dev** [comment]
No need, I can see it in the code now. I'll checkout your solution

**Review by kev-the-dev — APPROVED** [review]
I agree with this solution. Only downside is that some messages could be skipped if the callback is called before initializeStatisticsLogger() runs. Would appreciate comments above the new functions / variables describing what they do.

**Comment by DieHertz** [comment]
I have reworked it a bit in order to simplify, looks like two separate `init*` functions for `StatisticsLogger` aren't necessary.
Also added comments where appropriate.

**Comment by dirk-thomas** [comment]
Please retarget the latest default branch (currently `melodic-devel`) to be considered for review and merge.

**Comment by dirk-thomas** [comment]
Closing due to no further activity.

**Pull request closed without merging**